python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py

```diff
@@ -0,0 +1,203 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from typing import Union
+
+import cudf
+import dask_cudf
+import cugraph
+import pylibcugraph
+
+
+class EXPERIMENTAL__BulkSampler:
+    start_col_name = "_START_"
+    batch_col_name = "_BATCH_"
+
+    def __init__(
+        self,
+        batch_size: int,
+        output_path: str,
+        graph,
+        saturation_level: int = 200_000,
+        batches_per_partition=100,
+        rank: int = 0,
+        **kwargs,
+    ):
+        """
+        Constructs a new BulkSampler
+
+        Parameters
+        ----------
+        batch_size: int
+            The size of each batch.
+        output_path: str
+            The directory where results will be stored.
+        graph: cugraph.Graph
+            The cugraph graph to operate upon.
+        saturation_level: int (optional, default=200,000)
+            The number of samples that can be made within a single call.
+        batches_per_partition: int (optional, default=100)
+            The number of batches outputted to a single parquet partition.
+        rank: int (optional, default=0)
+            The rank of this sampler.  Used to isolate this sampler from
+            others that may be running on other nodes.
+        kwargs: kwargs
+            Keyword arguments to be passed to the sampler (i.e. fanout).
+        """
+        self.__batch_size = batch_size
+        self.__output_path = output_path
+        self.__graph = graph
+        self.__saturation_level = saturation_level
+        self.__batches_per_partition = batches_per_partition
+        self.__rank = rank
+        self.__batches = None
+        self.__sample_call_args = kwargs
+
+    @property
+    def rank(self) -> int:
+        return self.__rank
+
+    @property
+    def saturation_level(self) -> int:
+        return self.__saturation_level
+
+    @property
+    def batch_size(self) -> int:
+        return self.__batch_size
+
+    @property
+    def batches_per_partition(self) -> int:
+        return self.__batches_per_partition
+
+    @property
+    def size(self) -> int:
+        if self.__batches is None:
+            return 0
+        else:
+            return len(self.__batches)
+
+    def add_batches(
+        self,
+        df: Union[cudf.DataFrame, dask_cudf.DataFrame],
+        start_col_name: str,
+        batch_col_name: str,
+    ) -> None:
+        """
+        Adds batches to this BulkSampler.
+
+        Parameters
+        ----------
+        df: cudf.DataFrame or dask_cudf.DataFrame
+            Contains columns for vertex ids, batch id
+        start_col_name: str
+            Name of the column containing the start vertices
+        batch_col_name: str
+            Name of the column containing the batch ids
+
+        Returns
+        -------
+        """
```

Comment:
An example in this docstring would be great mainly because it might be nice to see an example of the input DF, but I don't know if doctests will try to run and verify it and if you're prepared to have the working.  A FIXME somewhere would be good otherwise.

Reply:
added example